src/k8bit/content.md

```diff
@@ -0,0 +1,623 @@
+
+1. You listed all Pods and stored the `resourceVersion`.
+1. You started a long-lived connection with the API and asked for updates. Only the updates after the last `resourceVersion` are streamed.
+1. You keep a local dictionary with all the Pods that you've seen so far.
+1. You handled reconnections when the connection is (abruptly) terminated.
+
+If you were to extend the same logic to Service and Deployments or any other Kubernetes resource, you probably want to have a very similar code.
+
+It's a good idea to encapsulate the above logic in a library, so you don't have to keep reinventing the wheel every time you wish to track objects.
+
+That's what the Kubernetes community thought too.
+
+In Kubernetes, there's a code pattern called the [Shared Informer](https://gianarb.it/blog/kubernetes-shared-informer).
+
+A shared informer encapsulates:
+
+1. The initial request to retrieve a list of resources.
+1. A Watch API request that starts from the previous `resourceVersion`.
+1. An efficient cache mechanism to store the resources locally in a dictionary.
+1. Reconnections when the connection is lost
+
+You can find an example of the shared informer in several programming languages:
+
+- [Go](https://gianarb.it/blog/kubernetes-shared-informer)
+- [Javascript/Typescript](https://github.com/kubernetes-client/javascript/blob/ced85fc4783d95fd0e944f88d98a576e1f55746a/examples/typescript/informer/informer.ts)
+- [Java](https://developers.redhat.com/blog/2019/10/07/write-a-simple-kubernetes-operator-in-java-using-the-fabric8-kubernetes-client/)
+- [Python (in progress)](https://github.com/kubernetes-client/python/issues/868)
+- [C# (in progress)](https://github.com/kubernetes-client/csharp/pull/394)
+
+Using the official Javascript client library for Kubernetes you can refactor your code in less than 20 lines of code:
+
+```javascript|title=informer.js
+const listFn = () => listPodForAllNamespaces();
+const informer = makeInformer(kc, '/api/v1/pods', listFn);
+
+informer.on('add', (pod) => { console.log(`Added: ${pod.metadata!.name}`); });
+informer.on('update', (pod) => { console.log(`Updated: ${pod.metadata!.name}`); });
+informer.on('delete', (pod) => { console.log(`Deleted: ${pod.metadata!.name}`); });
+informer.on('error', (err) => {
+  console.error(err)
+  // Restart informer after 5sec
+  setTimeout(() => informer.start(), 5000)
+})
+
+informer.start()
+```
+
+> Please note that the above code works only on Node.js and can't be run in the browser at the moment. This is a limitation of the client library.
+
+## Summary
+
+All code written so far ran against `kubectl proxy`.
+
+However, the same code could be repackaged and deployed inside your cluster.
+
+Imagine being able to track Pods, Deployments, Services, DaemonSets, etc. from within a cluster.
+
+That's precisely what happens when you deploy an operator or a controller in Kubernetes.
+
+In a way, this article helped you to write (almost) your first Kubernetes operator.
```

Comment:
Not sure if it can be really called a controller or operator.

AFAIK, a controller watches for certain changes in the Kubernetes API and then acts on the Kubernetes API by creating, editing, or deleting resources (like the built-in controllers, e.g. Deployment Controller).

An operator is like a controller but watches for changes to a custom resource definition rather than to a built-in resource type.

The dashboard only watches for changes in the Kubernetes API, but doesn't act back on the API (it doesn't "control" anything). So, the dashboard is rather just a client or observer of the Kubernetes API.

Reply:
I think the two terms used to be well defined:

- controller: only API
- operator: API + CRDs

Now I can see them being used interchangeably. Also, the use of the word controller has gone down drastically, so I feel like I'm trapped:

- It's a controller, but if I say so, people will wonder what a controller is in Kubernetes.
- If I say operator, it's technically incorrect, but the reader will understand the meaning.

I might just not use any of the above an explain what I mean. 🤔

Reply:
I wouldn't use any of the words because it just contributes to the confusion that there already is.